addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/ThirdGeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,767 @@
+             */
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    COMMUNICATION_ERROR_AUTHENTICATION);
+            return;
+        }
+        try {
+            data = cipher.doFinal(token.getBytes("UTF-8"));
+        } catch (IllegalBlockSizeException | BadPaddingException | UnsupportedEncodingException e1) {
+            /*
+             * No JSON answer received
+             */
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, COMMUNICATION_ERROR_JSON);
+            return;
+        }
+
+        byte[] ciphertext = new byte[data.length - AES_GCM_TAG_LENGTH / 8];
+        byte[] gcmTag = new byte[AES_GCM_TAG_LENGTH / 8];
+        System.arraycopy(data, 0, ciphertext, 0, data.length - AES_GCM_TAG_LENGTH / 8);
+        System.arraycopy(data, data.length - AES_GCM_TAG_LENGTH / 8, gcmTag, 0, AES_GCM_TAG_LENGTH / 8);
+
+        JsonObject createSessionJsonObject = new JsonObject();
+        createSessionJsonObject.addProperty("transactionId", transactionId);
+        createSessionJsonObject.addProperty("iv", Base64.getEncoder().encodeToString(iv));
+        createSessionJsonObject.addProperty("tag", Base64.getEncoder().encodeToString(gcmTag));
+        createSessionJsonObject.addProperty("payload", Base64.getEncoder().encodeToString(ciphertext));
+
+        /*
+         * finally create the session for further communication
+         */
+        ContentResponse createSessionResponseContentResponse;
+        JsonObject createSessionResponseJsonObject;
+        try {
+            createSessionResponseContentResponse = executeHttpPost(AUTH_CREATE_SESSION, createSessionJsonObject);
+            createSessionResponseJsonObject = getJsonObjectFromResponse(createSessionResponseContentResponse);
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    COMMUNICATION_ERROR_AUTHENTICATION);
+            return;
+        }
+        /*
+         * 200 is the desired status code
+         */
+        if (createSessionResponseContentResponse.getStatus() == 400) {
+            /*
+             * Authentication failed
+             */
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    CONFIGURATION_ERROR_PASSWORD);
+            return;
+        }
+
+        sessionId = createSessionResponseJsonObject.get("sessionId").getAsString();
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    /**
+     * Helper function to execute a HTTP post request
+     *
+     * @param resource   web API resource to post to
+     * @param parameters the JSON content to post
+     * @return the HTTP response for the created post request
+     * @throws Exception thrown if there are communication problems
+     */
+    private ContentResponse executeHttpPost(String resource, JsonObject parameters) throws Exception {
```

Comment:
As a general design: try to put as much as possible code related to the communication in separate class(es) and restrict  this class as much as possible only to handling code related to thing status updating/handling.

Reply:
I moved the http stuff to a dedicated helper class

Reply:
And just after hitting the commit button, I also moved the crypto stuff to a helper class… this will be included in my next commit… (should start thinking before pushing Buttons... :-) )

Reply:
Crypto stuff also implemented as helper class